python/ray/data/dataset.py

```diff
@@ -455,46 +457,41 @@ def map_batches(
         context = DatasetContext.get_current()
 
         def transform(
-            block: Block,
+            blocks: Iterable[Block],
             batch_fn: BatchUDF,
             *fn_args,
             **fn_kwargs,
         ) -> Iterable[Block]:
             DatasetContext._set_current(context)
             output_buffer = BlockOutputBuffer(None, context.target_max_block_size)
-            block = BlockAccessor.for_block(block)
-            total_rows = block.num_rows()
-            max_batch_size = batch_size
-            if max_batch_size is None:
-                max_batch_size = max(total_rows, 1)
-
-            for start in range(0, total_rows, max_batch_size):
-                # Build a block for each batch.
-                end = min(total_rows, start + max_batch_size)
-                # Make sure to copy if slicing to avoid the Arrow serialization
-                # bug where we include the entire base view on serialization.
-                view = block.slice(start, end, copy=batch_size is not None)
+            batcher = Batcher(batch_size)
```

Comment:
It was nice that the `Batcher` was usable here without any changes! A sign of a well-design abstraction. 😉